File: Loden Gendun/wk01 - starts 7th Mar/5-fri/mta.js
// SUBWAY object holds line arrays of stops.
const subway = {
    N: ['Times Square', '34th', '28th', '23rd', 'Union Square', '8th'],
    L: ['8th', '6th', 'Union Square', '3rd', '1st'],
    '6': ['Grand Central', '33rd', '28th', '23rd', 'Union Square', 'Astor Place']
};

// HELPER FUNCTIONS---------------------------------------------------------------------------
const changeRequired = function(lineAt, lineDest) {
    return lineAt != lineDest; // return true if commuter not on the line of their destination
};

// Determines whether iterator increments or decrements
const direction = function(commuter, nextStop) {
    const line = commuter.line;
    if (subway[line].indexOf(commuter.position) < subway[line].indexOf(nextStop)) {
        return 1;
    };
    return -1;
};

// Determines whether commuter has reached their next stop.
const tripOver = function(i, commuter, nextStop) {
    const line = commuter.line;
    if (direction(commuter, nextStop) > 0) {
         return i > subway[line].indexOf(nextStop);
    }
    return i < subway[line].indexOf(nextStop);
};

// Determines if commuter has reached their destination.
let destinationReached = function(commuter, lineOff, stopOff) {
    return commuter.line === lineOff && commuter.position === stopOff;
};
//--------------------------------------------------------------------------------------------

// MAIN FUNCTION--------------------------------------------------------------------------------------------------
const planTrip = function(lineOn, stopOn, lineOff, stopOff) {
    
    // Variables
    let changeOccured = false;
    let stopCount = 0;
    let stopsToTravel;
    let nextStop;
    
    // Keeps tracks of commuters line and position.
    let commuter = { 
        line: lineOn,
        position: stopOn
    };

    // Travel function moves commuter along train line to their next stop. 
    const travel = function(commuter, nextStop) {
        stopsToTravel = []; // Set new trip
    
        const line = commuter.line;
        const startPoint = subway[line].indexOf(commuter.position) + direction(commuter, nextStop)

        for (let i = startPoint; !tripOver(i, commuter, nextStop); i = i + direction(commuter, nextStop)) {
            commuter.position = subway[line][i];
            stopsToTravel.push(commuter.position);
        };

        stopCount += stopsToTravel.length; // Add stops travelled to stopCount.

        stopsToTravel = stopsToTravel.join(', '); // Convert stopsToTravel array to string for output.
    };
    
    // check if destination is on another line
    if (changeRequired(lineOn, lineOff)) {
        nextStop = 'Union Square';
    } else {
        nextStop = stopOff;
    };

    // Travel until commuter has reached their destination.
    while(!destinationReached(commuter, lineOff, stopOff)) {

        // Moves commuter from current position to their next stop.
        travel(commuter, nextStop);
        
        if (!changeOccured) {
            const output = `You must travel through the following stops on the ${lineOn} line: ${stopsToTravel}.`;
            console.log(output);
            if (changeRequired(lineOn, lineOff)) {
                console.log("Change at Union Square.");
            };
        } else {
            const output = `Your journey continues through the following stops: ${stopsToTravel}.`;
            console.log(output);
        };
        
        // Conditional for changing lines.
        if(commuter.line != lineOff) {
            commuter.line = lineOff;
            changeOccured = true;
        };
        if (nextStop != stopOff) {
            nextStop = stopOff;
        };
    };
    console.log(`${stopCount} stops in total.`);
};
//----------------------------------------------------------------------------------------------------------------

File: Loden Gendun/wk04 - starts 28th Mar/2-tue/mta.rb
$subway = { 'N' => ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"], 'L' => ["8th", "6th", "Union Square", "3rd", "1st"], '6' => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"] }

def travel(current_line, stop_on, next_stop) 
    line_indexes = [] << $subway[current_line].find_index(stop_on) << next_stop_index = $subway[current_line].find_index(next_stop)

    trip = $subway[current_line][line_indexes.min..line_indexes.max]

    trip.reverse! unless line_indexes == line_indexes.sort
    return trip
end

def plan_trip(line_on, stop_on, line_off, stop_off)
    stop_count = 0

    next_stop = stop_off

    next_stop = "Union Square" unless line_on == line_off #check if change required

    trip = travel line_on, stop_on, next_stop

    puts "You must travel through the following stops on the #{ line_on } line: #{ trip.drop(1).join ", " }."

    stop_count += trip.size - 1

    if line_on == line_off
        puts "#{ stop_count } stops in total"
        return true
    end
    
    puts "Change at Union Square."

    stop_on = "Union Square"

    current_line = line_off
    next_stop = stop_off
    
    trip = travel line_off, "Union Square", stop_off

    puts "Your journey continues through the following stops: #{ trip.drop(1).join ", " }."

    puts stop_count += trip.size - 1
end

binding.irb
